custom_components/tahoma/binary_sensor.py

```diff
@@ -66,21 +87,25 @@ def update(self):
 
         self.controller.get_states([self.tahoma_device])
 
-        if CORE_CONTACT_STATE in self.tahoma_device.active_states:
-            self.current_value = (
-                self.tahoma_device.active_states.get(CORE_CONTACT_STATE) == "open"
-            )
-
-        if CORE_OCCUPANCY_STATE in self.tahoma_device.active_states:
-            self.current_value = (
-                self.tahoma_device.active_states.get(CORE_OCCUPANCY_STATE)
-                == "personInside"
-            )
-
-        if CORE_SMOKE_STATE in self.tahoma_device.active_states:
-            self.current_value = (
-                self.tahoma_device.active_states.get(CORE_SMOKE_STATE) == "detected"
-            )
+        states = self.tahoma_device.active_states
+
+        if CORE_CONTACT_STATE in states:
+            self.current_value = states.get(CORE_CONTACT_STATE) == "open"
+
+        if CORE_OCCUPANCY_STATE in states:
+            self.current_value = states.get(CORE_OCCUPANCY_STATE) == "personInside"
+
+        if CORE_SMOKE_STATE in states:
+            self.current_value = states.get(CORE_SMOKE_STATE) == "detected"
+
+        if CORE_RAIN_STATE in states:
+            self.current_value = states.get(CORE_RAIN_STATE) == "detected"
+
+        if CORE_WATER_DETECTION_STATE in states:
+            self.current_value = states.get(CORE_WATER_DETECTION_STATE) == "detected"
+
+        if IO_VIBRATION_STATE in states:
+            self.current_value = states.get(IO_VIBRATION_STATE) == "detected"
```

Comment:
Should we have constants in the beginning of the file or in `const.py` instead of hardcoded words here for the ones not in `core/const.py`?

Reply:
I am not sure yet, to be honest. I like to have at least everything defined as a const, also the "detected" strings at some point. (future PR). However not sure if we should try to have the constants in the platform files and only shared constants in `consts.py`, or if we should keep every constant in a single file.

@tetienne do you have an opinion on this?

Reply:
ok. Let's not keep this from merging for now. We can make an enhancement PR later regarding constants

Reply:
It really depends. If the keyword "detected" is only related to the binary_sensor platform, put the constant here, else in constants.py.

Reply:
@tetienne @vlebourl lets do that in a future PR indeed. There are many constants in const.py that are only used for a single platform at the moment, thus this will be some work :O

Reply:
Do you know if the devices related to the "detected" cases have something in common? For instance the widget?

Reply:
@tetienne no, unfortunately not. That's why we need a mapping for Sensors to understand if it is a `binary_sensor` or a `sensor`. However at the moment, we support almost all available sensors.

Reply:
However, there are a few files with a lot of if statements now, that should be mapped using some helper functions. There has been some progress in cover for the commands, but it is something that could be done for states as well.

Reply:
https://github.com/iMicknl/ha-tahoma/pull/91/files <-- we should add a helper function like this at some point, for states and commands.

Reply:
Indeed this pattern occurred a lot.

Reply:
> @tetienne no, unfortunately not. That's why we need a mapping for Sensors to understand if it is a `binary_sensor` or a `sensor`. However at the moment, we support almost all available sensors.

Can you give me a device which has this CORE_WATER_DETECTION_STATE state and other which has CORE_GAS_DETECTION_STATE?

Reply:
`rtds:RTDSGasSensor` and `rtds:RTDSWaterSensor`

Reply:
> Indeed this pattern occurred a lot.

This is mainly due to the fact that it started small, but at the moment many devices are supported, thus asking for some refactors throughout the component.

Reply:
Logged an issue here; https://github.com/iMicknl/ha-tahoma/issues/110